src/libsyntax/codemap.rs

```diff
@@ -807,6 +807,70 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    pub fn span_to_expanded_string(&self, sp: Span) -> String {
+        let src_str = self.span_to_string(sp);
+        let src_snip = self.span_to_snippet(sp).unwrap_or(
+            "Snippet unavailable (span malformed or source missing)".to_string());
+        if sp.expn_id == NO_EXPANSION || sp.expn_id == COMMAND_LINE_EXPN {
+            return format!("{}\n{}", src_snip, src_str);
+        }
+
+        let mut strings = vec![src_str.into_boxed_str(), src_snip.into_boxed_str()];
+        let mut cur_span = sp;
+        while cur_span.expn_id != NO_EXPANSION && cur_span.expn_id != COMMAND_LINE_EXPN {
+            let old_span = cur_span;
+            let new_span = self.with_expn_info(old_span.expn_id,
+                                              |ei| ei.map(|ei| ei.call_site));
```

Comment:
You expand iteratively down the `call_site` branch, but only one step down the `callee` branch (below), it seems to me you should be iterating to the end of both branches (which would be easier if you were being recursive). At that point you probably need to think about indentation or something to distinguish the structure.